bob/src/cluster/quorum.rs

```diff
@@ -351,25 +352,194 @@ impl Cluster for Quorum {
+                        alien_index_map.update_existence(&mut result, remote_alien_result.inner());
+                    }
+                    Err(e) => {
+                        debug!("Failed to check existence in aliens: {:?}", e);
+                    }
                 }
             }
         }
-        Ok(exist)
+
+        Ok(result)
     }
 
     async fn delete(&self, key: BobKey, meta: &BobMeta) -> Result<(), Error> {
         self.delete_on_nodes(key, meta).await
     }
 }
+
+async fn collect_remote_exists(
+    result: &mut [bool],
+    keys: &[BobKey],
+    indexes_by_node: &mut HashMap<Node, IndexMap>,
+) -> Result<(), Error> {
+    if !indexes_by_node.is_empty() {
+        let mut keys_by_node = HashMap::new();
+        for (node, node_map) in indexes_by_node.iter_mut() {
+            node_map.retain_not_existed(&result);
+            if !node_map.is_empty() {
+                keys_by_node.insert(node.clone(), node_map.collect(keys));
+            }
+        }
+
+        if !keys_by_node.is_empty() {
+            let nodes: Vec<_> = keys_by_node.keys().cloned().collect();
+            let remote_results = exist_on_remote_nodes(&nodes, keys_by_node).await;
+            for (remote_result, node) in remote_results.into_iter().zip(nodes) {
+                match remote_result {
+                    Ok(remote_result) => {
+                        indexes_by_node
+                            .get(&node)
+                            .map(|idx| idx.update_existence(result, remote_result.inner()));
+                    }
+                    Err(e) => {
+                        debug!("Failed to check existence on node {}: {:?}", node.name(), e);
+                    }
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+fn group_by_nodes(
+    keys: &[BobKey],
+    mapper: &Virtual,
+) -> (
+    Option<IndexMap>,
+    HashMap<Node, IndexMap>,
+    HashMap<Node, IndexMap>,
+) {
+    let mut local = IndexMap::new();
+    let mut primary = HashMap::new();
+    let mut secondary = HashMap::new();
+
+    let local_node = mapper.local_node_name();
+
+    for (index, &key) in keys.iter().enumerate() {
+        let target_nodes = mapper.get_target_nodes_for_key(key);
+
+        if !target_nodes
+            .iter()
+            .any(|n| n.name() == local_node || primary.contains_key(n))
+        {
+            if let Some(node) = target_nodes.iter().find(|n| !secondary.contains_key(*n)) {
+                primary.insert(node.clone(), IndexMap::new());
+            }
+        }
+
+        for node in target_nodes {
+            if node.name() == local_node {
+                local.push(index);
+            } else if let Some(map) = primary.get_mut(node) {
+                map.push(index)
+            } else {
+                secondary
+                    .entry(node.clone())
+                    .or_insert(IndexMap::new())
+                    .push(index);
+            }
+        }
+    }
+
+    (
```

Comment:
Don't like this syntax. Please, add `return` keyword to emphasis, that this is a result of a function